File: 09-语言学习/Golang/内存模型与sync原语/python/memory_model.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""Go 内存模型：happens-before 关系的可执行模型。

依据 https://go.dev/ref/mem（2022-06-06 版）原文：
  * 内存操作分 read-like（read / atomic read / mutex lock / channel receive）与
    write-like（write / atomic write / mutex unlock / channel send / channel close），
    另有一类 **synchronizing operation**（原子访问、mutex 操作、channel 操作）。
  * `synchronized before` 由映射 W 导出：若同步读 r 观察到同步写 w，则 w synchronized before r。
  * `happens before` = `sequenced before` ∪ `synchronized before` 的**传递闭包**。
  * 数据竞争 = 同一内存位置上两个操作、至少一个非同步、且在 happens-before 下不可比。
  * **DRF-SC**：无数据竞争的程序，其行为等价于各 goroutine 顺序执行的某个交错。
被 python/main.py 以 `from memory_model import *` 使用。
"""

READ_LIKE = ("read", "atomic_read", "lock", "recv")
WRITE_LIKE = ("write", "atomic_write", "unlock", "send", "close")


class Op:
    __slots__ = ("gid", "label", "kind", "loc", "sync")

    def __init__(self, gid, label, kind, loc=None):
        self.gid = gid          # 所在 goroutine
        self.label = label      # 人类可读名字
        self.kind = kind        # read / write / send / recv / close / lock / unlock ...
        self.loc = loc          # 访问的内存位置（None 表示不访问内存）
        self.sync = kind in READ_LIKE + WRITE_LIKE and kind not in ("read", "write")

    @property
    def read_like(self):
        return self.kind in READ_LIKE

    @property
    def write_like(self):
        return self.kind in WRITE_LIKE

    def __repr__(self):
        return self.label


class Execution:
    """一次程序执行：若干 goroutine 的操作序列 + 同步边。"""

    def __init__(self):
        self.ops = []           # 按加入顺序，即 sequenced before 的近似（同 goroutine 即成立）
        self.sync_edges = []    # (a, b)：a synchronized before b
        self.notes = []

    def op(self, gid, label, kind, loc=None):
        o = Op(gid, label, kind, loc)
        self.ops.append(o)
        return o

    def sync(self, a, b, why=""):
        self.sync_edges.append((a, b))
        if why:
            self.notes.append("%s  synchronized before  %s   （%s）" % (a.label, b.label, why))
        return self

    # -------------------------------------------------- 关系计算
    def sequenced_before(self, a, b):
        """同一 goroutine 内、按程序顺序先出现。"""
        return a.gid == b.gid and a is not b and self.ops.index(a) < self.ops.index(b)

    def happens_before(self):
        """sequenced ∪ synchronized 的传递闭包（Floyd–Warshall 于操作集合上）。"""
        n = len(self.ops)
        idx = {id(o): i for i, o in enumerate(self.ops)}
        hb = [[False] * n for _ in range(n)]
        for i in range(n):
            for j in range(n):
                if i != j and self.sequenced_before(self.ops[i], self.ops[j]):
                    hb[i][j] = True
        for a, b in self.sync_edges:
            hb[idx[id(a)]][idx[id(b)]] = True
        for k in range(n):
            for i in range(n):
                if hb[i][k]:
                    row_k = hb[k]
                    row_i = hb[i]
                    for j in range(n):
                        if row_k[j]:
                            row_i[j] = True
        return hb, idx

    def ordered(self, a, b):
        hb, idx = self.happens_before()
        i, j = idx[id(a)], idx[id(b)]
        return hb[i][j] or hb[j][i]

    def is_race(self, a, b):
        """同一位置、至少一个非同步、且 HB 不可比 → 数据竞争。"""
        if a.loc is None or a.loc != b.loc:
            return False
        if not (a.read_like or a.write_like) or not (b.read_like or b.write_like):
            return False
        if a.read_like and b.read_like:
            return False                     # 读-读不构成竞争
        if a.sync and b.sync:
            return False                     # 两个同步操作之间不算竞争
        return not self.ordered(a, b)

    def races(self):
        out = []
        for i, a in enumerate(self.ops):
            for b in self.ops[i + 1:]:
                if self.is_race(a, b):
                    out.append((a.label, b.label))
        return out

    def visible_writes(self, rd):
        """按 Requirement 3：可见写 = 发生在 r 之前、且没有被别的发生在 r 之前的写覆盖。"""
        hb, idx = self.happens_before()
        i = idx[id(rd)]
        cands = [self.ops[k] for k in range(len(self.ops))
                 if hb[k][i] and self.ops[k].write_like and self.ops[k].loc == rd.loc]
        vis = []
        for w in cands:
            j = idx[id(w)]
            if not any(hb[j][idx[id(w2)]] for w2 in cands if w2 is not w):
                vis.append(w)
        return vis

File: 09-语言学习/Golang/内存模型与sync原语/go/memory_model.go
// Go 内存模型：happens-before 关系的可执行模型（与 python/memory_model.py 同构）。
//
// 依据 https://go.dev/ref/mem（2022-06-06）：synchronized before 由映射 W 导出，
// happens before = sequenced ∪ synchronized 的传递闭包，数据竞争 = 同位置、至少一个
// 非同步、且在 happens-before 下不可比；DRF-SC = 无竞争程序等价于某顺序交错。
package main

var readLike = map[string]bool{
	"read": true, "atomic_read": true, "lock": true, "recv": true,
}

var writeLike = map[string]bool{
	"write": true, "atomic_write": true, "unlock": true, "send": true, "close": true,
}

type op struct {
	gid   int
	label string
	kind  string
	loc   string
}

func (o *op) isRead() bool  { return readLike[o.kind] }
func (o *op) isWrite() bool { return writeLike[o.kind] }

// isSync 只认「同步操作」：普通 read / write 不算同步操作，
// 因此「两个同步操作之间不算竞争」这条豁免不能把普通读写也豁免掉。
func (o *op) isSync() bool {
	if o.kind == "read" || o.kind == "write" {
		return false
	}
	return o.isRead() || o.isWrite()
}

type execution struct {
	ops   []*op
	syncs [][2]*op
}

func (e *execution) add(gid int, label, kind, loc string) *op {
	o := &op{gid: gid, label: label, kind: kind, loc: loc}
	e.ops = append(e.ops, o)
	return o
}

func (e *execution) sync(a, b *op) *execution {
	e.syncs = append(e.syncs, [2]*op{a, b})
	return e
}

func (e *execution) index(o *op) int {
	for i, x := range e.ops {
		if x == o {
			return i
		}
	}
	return -1
}

// happensBefore 返回传递闭包矩阵与下标映射。
func (e *execution) happensBefore() ([][]bool, map[*op]int) {
	n := len(e.ops)
	idx := map[*op]int{}
	for i, o := range e.ops {
		idx[o] = i
	}
	hb := make([][]bool, n)
	for i := range hb {
		hb[i] = make([]bool, n)
	}
	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			a, b := e.ops[i], e.ops[j]
			if i != j && a.gid == b.gid && i < j {
				hb[i][j] = true // sequenced before
			}
		}
	}
	for _, s := range e.syncs {
		hb[idx[s[0]]][idx[s[1]]] = true
	}
	for k := 0; k < n; k++ {
		for i := 0; i < n; i++ {
			if !hb[i][k] {
				continue
			}
			for j := 0; j < n; j++ {
				if hb[k][j] {
					hb[i][j] = true
				}
			}
		}
	}
	return hb, idx
}

func (e *execution) ordered(a, b *op) bool {
	hb, idx := e.happensBefore()
	i, j := idx[a], idx[b]
	return hb[i][j] || hb[j][i]
}

func (e *execution) isRace(a, b *op) bool {
	if a.loc == "" || a.loc != b.loc {
		return false
	}
	if !(a.isRead() || a.isWrite()) || !(b.isRead() || b.isWrite()) {
		return false
	}
	if a.isRead() && b.isRead() {
		return false // 读-读不构成竞争
	}
	if a.isSync() && b.isSync() {
		return false
	}
	return !e.ordered(a, b)
}

func (e *execution) races() [][2]string {
	out := [][2]string{}
	for i, a := range e.ops {
		for _, b := range e.ops[i+1:] {
			if e.isRace(a, b) {
				out = append(out, [2]string{a.label, b.label})
			}
		}
	}
	return out
}

// visibleWrites 按 Requirement 3：发生在 r 之前、且没有被其他「也发生在 r 之前」的写覆盖。
func (e *execution) visibleWrites(rd *op) []*op {
	hb, idx := e.happensBefore()
	i := idx[rd]
	cands := []*op{}
	for k, o := range e.ops {
		if hb[k][i] && o.isWrite() && o.loc == rd.loc {
			cands = append(cands, o)
		}
	}
	vis := []*op{}
	for _, w := range cands {
		covered := false
		for _, w2 := range cands {
			if w2 != w && hb[idx[w]][idx[w2]] {
				covered = true
			}
		}
		if !covered {
			vis = append(vis, w)
		}
	}
	return vis
}
